etcdserver/security/security.go

```diff
@@ -0,0 +1,303 @@
+}
+
+func (s *Store) GetUser(name string) (User, error) {
+	// TODO(barakmich): Possibly add a cache
+	resp, err := s.requestResource("/users/"+name, false)
+	if err != nil {
+		return User{}, err
+	}
+	var u User
+	err = json.Unmarshal([]byte(*resp.Event.Node.Value), &u)
+	if err != nil {
+		return u, err
+	}
+
+	return u, nil
+}
+
+func (s *Store) CreateUser(user User) error {
+	hash, err := bcrypt.GenerateFromPassword([]byte(user.Password), bcrypt.DefaultCost)
+	if err != nil {
+		return err
+	}
+	user.Password = string(hash)
+	user.Modified = time.Now()
+
+	_, err = s.createResource("/users/"+user.User, user)
+	return err
+}
+
+func (s *Store) DeleteUser(name string) error {
+	_, err := s.deleteResource("/users/" + name)
+	return err
+}
+
+func (s *Store) UpdateUser(user User) (User, error) {
+	old, err := s.GetUser(user.User)
+	if err != nil {
+		return old, err
+	}
+	newUser, err := old.Merge(user)
+	if err != nil {
+		return old, err
+	}
+	_, err = s.updateResource("/users/"+user.User, newUser)
+	if err != nil {
+		return newUser, err
+	}
+	return newUser, nil
+}
+
+func (s *Store) AllRoles() ([]string, error) {
+	resp, err := s.requestResource("/roles/", false)
+	if err != nil {
+		return nil, err
+	}
+	var nodes []string
+	nodes = append(nodes, "root")
+	for _, n := range resp.Event.Node.Nodes {
+		_, role := path.Split(n.Key)
+		nodes = append(nodes, role)
+	}
+	sort.Strings(nodes)
+	return nodes, nil
+}
+
+func (s *Store) GetRole(name string) (Role, error) {
+	// TODO(barakmich): Possibly add a cache
+	resp, err := s.requestResource("/roles/"+name, false)
+	if err != nil {
+		return Role{}, err
+	}
+	var r Role
+	err = json.Unmarshal([]byte(*resp.Event.Node.Value), &r)
+	if err != nil {
+		return r, err
+	}
+
+	return r, nil
+}
+
+func (s *Store) CreateRole(role Role) error {
+	role.Modified = time.Now()
+	_, err := s.createResource("/roles/"+role.Role, role)
+	return err
+}
+
+func (s *Store) DeleteRole(name string) error {
+	_, err := s.deleteResource("/roles/" + name)
+	return err
+}
+
+func (s *Store) UpdateRole(role Role) (Role, error) {
+	old, err := s.GetRole(role.Role)
+	if err != nil {
+		return old, err
+	}
+	newRole, err := old.Merge(role)
+	if err != nil {
+		return old, err
+	}
+	_, err = s.updateResource("/roles/"+role.Role, newRole)
+	if err != nil {
+		return newRole, err
+	}
+	return newRole, nil
+}
+
+func (u User) Merge(n User) (User, error) {
+	var out User
+	if u.User != n.User {
+		return out, mergeErr("Merging user data with conflicting usernames: %s %s", u.User, n.User)
+	}
+	out.User = u.User
+	if n.Password != "" {
+		hash, err := bcrypt.GenerateFromPassword([]byte(n.Password), bcrypt.DefaultCost)
+		if err != nil {
+			return out, err
+		}
+		out.Password = string(hash)
+	} else {
+		out.Password = u.Password
+	}
+	currentRoles := types.NewUnsafeSet(u.Roles...)
+	for _, g := range n.Grant {
+		if currentRoles.Contains(g) {
+			return out, mergeErr("Granting duplicate role %s for user %s", g, n.User)
```

Comment:
I figure I'd take the most restrictive notions possible. It could be an error. Ditto below.